java/code/src/com/redhat/rhn/frontend/servlets/AuthorizationFilter.java

```diff
@@ -0,0 +1,198 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import spark.route.HttpMethod;
+import spark.route.ServletRoutes;
+import spark.routematch.RouteMatch;
+
+/**
+ * A servlet filter to enforce access control rules
+ */
+public class AuthorizationFilter implements Filter {
+
+    private static final Logger LOG = LogManager.getLogger(AuthorizationFilter.class);
+    private AuthenticationService authenticationService;
+
+    @Override
+    public void init(FilterConfig config) {
+        AuthenticationServiceFactory factory = AuthenticationServiceFactory.getInstance();
+        authenticationService = factory.getAuthenticationService();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException,
+            ServletException {
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("ENTER AuthorizationFilter.doFilter: {} [{}] ({})", request.getRemoteAddr(), new Date(),
+                    ((HttpServletRequest) (request)).getRequestURI());
+        }
+
+        HttpServletRequest hreq = new RhnHttpServletRequest((HttpServletRequest) request);
+        HttpServletResponse hres = (HttpServletResponse) response;
+        Set<String> noAuthEndpoints = WebEndpointFactory.getUnauthorizedWebEndpoints();
+
+        try {
+            if (hreq.getServletPath().startsWith("/ajax")) {
+                handleAjaxAccess(hreq, noAuthEndpoints);
+            }
+            else if (hreq.getServletPath().endsWith(".do") || hreq.getServletPath().endsWith(".jsp")) {
+                handleStrutsAccess(hreq, noAuthEndpoints);
+            }
+            else {
+                handleSparkAccess(hreq, noAuthEndpoints);
+            }
+            LOG.debug("Access granted for user '{}' to URI '{}' [{}]",
+                    new RequestContext(hreq).getCurrentUser(), hreq.getRequestURI(), hreq.getMethod());
+        }
+        catch (PermissionException e) {
+            // TODO: Handle PermissionExceptions properly depending on the "Content-Type" and "Accept" headers
+            // TODO: possibly pass it down another filter
+            // TODO: Review PageFilter
+            LOG.debug("Access restricted for user '{}' to URI '{}' [{}]",
+                    new RequestContext(hreq).getCurrentUser(), hreq.getRequestURI(), hreq.getMethod());
+            if (ConfigDefaults.get().isRbacEnabled()) {
+                hres.sendError(HttpServletResponse.SC_FORBIDDEN, e.getMessage());
+                return;
+            }
+        }
+
+        // Pass control on to the next filter
+        chain.doFilter(request, response);
+    }
+
+    private void handleAjaxAccess(HttpServletRequest hreq, Set<String> noAuthEndpoints) {
+        String path = hreq.getServletPath() + hreq.getPathInfo();
+        if (!authenticationService.requestURIRequiresAuthentication(hreq) || noAuthEndpoints.contains(path)) {
```

Comment:
I would prefer to not have a dependency from `AuthenticationServiceFactory`, but I was looking at the endpoints in there and they are super generic (like `/css`), so would be hard to move them to the database.